crates/sui-core/src/batch_bls_verifier.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use shared_crypto::intent::{Intent, IntentScope};
+use sui_types::{
+    committee::Committee,
+    crypto::{AuthoritySignInfoTrait, VerificationObligation},
+    error::SuiResult,
+    message_envelope::Message,
+    messages::CertifiedTransaction,
+};
+
+use tap::TapFallible;
+use tracing::error;
+
+/// Verifies all certificates - if any fail return error.
+pub fn batch_verify_all_certificates(
+    committee: &Committee,
+    certs: &[CertifiedTransaction],
+) -> SuiResult {
+    // Verify user signatures
+    for cert in certs {
+        cert.data().verify(None)?;
+    }
+
+    batch_verify_certificates_impl(committee, certs)
+}
+
+/// Verifies certificates in batch mode, but returns a separate result for each cert.
+pub fn batch_verify_certificates(
+    committee: &Committee,
+    certs: &[CertifiedTransaction],
+) -> Vec<SuiResult> {
+    match batch_verify_certificates_impl(committee, certs) {
```

Comment:
- don't we need to check the user sigs in `Err(_)` as well?
- verifying user sigs is still cheaper than BLS batch verify per cert, so it's better to check them before calling `batch_verify_certificates_impl`

Reply:
re (b): it's not always true that user sig is lighter (ie when large multi-sig are used - especially with the current slow R1 impl), but with high probability it will indeed be amortized cheaper

Reply:
@benr-ml we do, and we are: `verify_signature` checks user signatures.

Reply:
also as far as verification order - we should never ever see a valid cert with invalid user sigs. And as far as attacks go, it is trivial to generate an invalid cert over a valid user sig, so we can't prevent an attacker from wasting cpu in this case. Given that, I just chose to go with the order that resulted in simpler code. Let me know if you disagree, I'm open to changing it.